include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,423 @@
+
+      // 2.3 Update of ideal point.
+      for (size_t idx = 0;idx < numObjectives;idx++)
+      {
+        idealPoint[0](idx) = std::min(idealPoint[0](idx),
+            evaluatedCandidate[0][idx]);
+      }
+
+      // 2.4 Update of the neighbouring solutions.
+      for (size_t idx = 0;idx < neighbourhoodSize;idx++)
+      {
+        if (DecomposedSingleObjective(weights[weightNeighbourIndices(i, idx)],
+              idealPoint[0], evaluatedCandidate[0])
+            <= DecomposedSingleObjective(
+              weights[weightNeighbourIndices(i,idx)],
+              idealPoint[0], FValue[weightNeighbourIndices(i, idx)]))
+        {
+          population.at(weightNeighbourIndices(i, idx)) = candidate[0];
+          FValue[weightNeighbourIndices(i, idx)] = evaluatedCandidate[0];
+        }
+      }
+
+      // 2.5 Updating External Population.
+      if ( !externalPopulation.empty() )
+      {
+        std::vector<arma::vec> first(1);
+        auto df = [&](MatType firstMat) -> bool
+        {
+          std::vector<MatType> wrapperFirst(1), wrapperSecond(1);
+          wrapperFirst[0] = firstMat;
+          first[0].resize(numObjectives);
+          EvaluateObjectives(wrapperFirst, objectives, first);
+          return Dominates(evaluatedCandidate[0], first[0]);
+        };
+        //! Remove the part that is dominated by candidate.
+        externalPopulation.erase(
+            std::remove_if(
+              externalPopulation.begin(),
+              externalPopulation.end(),
+              df), externalPopulation.end());
+
+        //! Check if any of the remaining members of external population dominate
+        //! candidate.
+        bool flag = 0;
+        std::vector<MatType> wrapperFirst(1);
+        for (size_t idx = 0; idx < externalPopulation.size(); idx++)
+        {
+          wrapperFirst[0]=externalPopulation[idx];
+          first[0].clear();
+          first[0].resize(numObjectives);
+          EvaluateObjectives(wrapperFirst, objectives, first);
+          if (Dominates(first[0], evaluatedCandidate[0]))
+          {
+            flag = 1;
+            break;
+          }
+        }
+        if (flag == 0)
+        {
+          externalPopulation.push_back(candidate[0]);
+          externalPopulationFValue.push_back(evaluatedCandidate[0]);
+        }
+      }
+      else
+      {
+        externalPopulation.push_back(candidate[0]);
+        externalPopulationFValue.push_back(evaluatedCandidate[0]);
+      }
+    }
+  }
+  bestFront = std::move(externalPopulation);
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+
+  ElemType performance = std::numeric_limits<ElemType>::max();
+
+  for (arma::Col<ElemType> objective: externalPopulationFValue)
+  {
+    if (arma::accu(objective) < performance)
+      performance = arma::accu(objective);
+  }
+
+  return performance;
+}
+
+//! Perform mutation of the candidate.
+template<typename MatType>
+inline void MOEAD::Mutate(MatType& child,
```

Comment:
Is this the Differential Evolution (DE) operator and polynomial mutation procedure as in Li and Zhang (2009)?

Reply:
Yes. Actually I used the implementation I had mentioned earlier to implement it, so I forgot to cite the paper.